### PR TITLE
Jkralik/deadlock

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: api/cloud/oc_cloud_manager.c [hunk 1]
```diff
@@ -263,8 +256,22 @@ static void
...
+    case CLOUD_OK:
+      cloud_set_cps_and_last_error(ctx, OC_CPS_REGISTERED, CLOUD_OK);
+      break;
+    case CLOUD_ERROR_CONNECT:
+      if (!is_retry_over(ctx)) {
+        // While retrying, keep last error (clec) to CLOUD_OK
+        cloud_set_cps_and_last_error(ctx, OC_CPS_REGISTERING, CLOUD_OK);
+        break;
+      }
+      /* FALLTHRU */
```
Comment (Danielius1922) [inline api/cloud/oc_cloud_manager.c:270]: typo: fallthrough

File: api/cloud/oc_cloud_manager.c [hunk 2]
```diff
@@ -223,33 +222,27 @@ _register_handler(oc_cloud_context_t *ctx, oc_client_response_t *data)
...
 }
 
 void
 oc_cloud_register_handler(oc_client_response_t *data)
 {
   cloud_api_param_t *p = (cloud_api_param_t *)data->user_data;
   oc_cloud_context_t *ctx = p->ctx;
-  _register_handler(ctx, data);
+  oc_cloud_error_t err = _register_handler(ctx, data);
+  switch (err) {
```
Comment (Danielius1922) [inline api/cloud/oc_cloud_manager.c:239]: Nitpick: The `switch` works, though `CLOUD_OK ` and `default`, alternative would be to add `bool` parameter to `_register_handler` to control whether `retry` should be run. However, 7 duplicated lines is not something we should lose sleep over, so I don't mind if it stays like this.

File: api/cloud/oc_cloud_manager.c [hunk 3]
```diff
@@ -340,6 +347,17 @@ _login_handler(oc_cloud_context_t *ctx, oc_client_response_t *data)
     goto error;
   }
 
+  int64_t expires_in = 0;
+  if (!oc_rep_get_int(data->payload, EXPIRESIN_KEY, &expires_in)) {
+    err = CLOUD_ERROR_RESPONSE;
+    goto error;
+  }
+  ctx->store.expires_in = expires_in;
+  if (ctx->store.expires_in > 0) {
```
Comment (Danielius1922) [inline api/cloud/oc_cloud_manager.c:356]: line 365?

Comment (jkralik) [reply api/cloud/oc_cloud_manager.c:347]: Fixed

File: api/cloud/oc_cloud_manager.c (hunk 1 above)
Comment (jkralik) [reply api/cloud/oc_cloud_manager.c]: Removed

File: api/cloud/oc_cloud_manager.c (hunk 2 above)
Comment (jkralik) [reply api/cloud/oc_cloud_manager.c]: I used the `retry` flag

**Review by Danielius1922 — APPROVED** [review]

**Pull request merged**